### DISCUSSION TITLE
[Help]: Is there a way to create 2 fully synchronized carousels?

### DISCUSSION THREAD

**Discussion by 7iomka** [davidjerleke/embla-carousel#567] [Help]
I tried this
```tsx
 const [selectedIndex, setSelectedIndex] = useState(0);
  const [scrollProgress, setScrollProgress] = useState(0);

  const [embla, setEmbla] = useState<EmblaCarouselType | null>(null);
  const [emblaReflection, setEmblaReflection] = useState<EmblaCarouselType | null>(null);

  const handleScroll = useCallback(() => {
    if (!embla || !emblaReflection) return;
    const progress = Math.max(0, Math.min(1, embla.scrollProgress()));
    setScrollProgress(progress * 100);
    emblaReflection.scrollTo(embla.selectedScrollSnap());
  }, [embla, emblaReflection, setScrollProgress]);

  const handleSelect = useCallback(() => {
    if (!embla || !emblaReflection) return;
    setSelectedIndex(embla.selectedScrollSnap());
    emblaReflection.scrollTo(embla.selectedScrollSnap());
  }, [embla, emblaReflection, setSelectedIndex]);


  useEffect(() => {
    if (!embla) return;
    handleSelect();
    embla.on('select', handleSelect);

    embla.on('scroll', handleScroll);
    handleScroll();

    return () => {
      embla.off('select', handleSelect);
      embla.off('scroll', handleScroll);
    };
  }, [embla, handleSelect, handleScroll]);
```

So, in this case when I scroll the first (`embla`) carousel by hand or somehow else - I want to have the same scroll progress in the second carousel (`emblaReflection`).
Of course, I tried to store scrollProgress in the state, but in v8 I don't found any method to set carousel progress.
So, currently the 2nd carousel just changes active slide, when the first one has changed it, and this is not that I expect from fully synchronized carousels.

Please, help me with this trivial task.

**Comment by davidjerleke** [comment]
Hi @7iomka,

Do you want carousel `A` to follow carousel `B`, or do you want both carousel `A` to follow `B` and `B` to follow `A`?

Best,
David

**Comment by 7iomka** [comment]
Hi @davidjerleke 
I would like to see the the second case, because I want full synchronization between both carousels and this example would be more common that the first one: Independent to what carousel is triggered (scrolling left/right, change slide by navigation or somehow else), both needed to have the same state.
In my realword case I have layout with
- image
- image reflection
- thumbnails
User can can scroll image carousel, and need to see that image reflection carousel is scrolled by exactly the same value and same way immediately when this change occurs (because it is reflection :-) )
Thumbnails it is just list with length of slides count, with onclick handler, that change index of one of available carousels.
Currently image reflection carousel is not accessible by design, but if this case occurs in the future with 2 accessible carousels, I need to know how to also cross-sync it.

Thanks!

**Reply by davidjerleke** [reply]
Thanks for the additional details @7iomka. I also want to know if carousel `A` will be equal to carousel `B` in terms of width, number of slides and all other aspects?

**Comment by 7iomka** [comment]
> Thanks for the additional details @7iomka. I also want to know if carousel `A` will be equal to carousel `B` in terms of width, number of slides and all other aspects?

As for the number of elements in the carousel - it is required for synchronization - that it should be the same for synchronized carousels.
About slide sizes -
the main idea is to keep the scroll percentage without depending on static values.
In my particular case, the carousel with reflection has the same container size as the main carousel.
In other cases this may not be a requirement.

**Reply by 7iomka** [reply]
@davidjerleke Hi)
Did you manage to make an example according to my wishes or is it impossible to do it in the current implementation)?

**Reply by davidjerleke** [reply]
@7iomka it’s definitely possible to achieve with the help of `emblaApi.internalEngine()` method but I simply haven’t had the time to look into it.

I’m sorry if this makes you disappointed but I can’t tell you when I will have the time to look into this. I’m the sole developer on this project and I have a ton to do before the stable v8 release. This is a special request that isn’t part of the public API and therefore the priority is low. 

Anyone who wants help with a special request fast can’t expect me to spend even more of my unpaid spare time than I already do on this project. The only way I will help someone out fast with their special request is if they make an agreement with me and pay me for my time.

Thank you for understanding.

Best,
David

**Comment by SemStassen** [comment]
Hi, quickly chiming in here since I'm also interested in the same use case.

After having a quick look at `emblaApi.internalEngine()` I'm not sure how the sync can be achieved, since the `scrollProgress()` method only seems to have a getter.

I understand you are busy, and I'm willing to look into this more myself, but was hoping you could point me in the right direction as I am not that familiar with this package.

Thanks!

**Reply by davidjerleke** [reply]
@SemStassen could you specify what you want to achieve by answering [this question](https://github.com/davidjerleke/embla-carousel/discussions/567#discussioncomment-6985456)? Also, are you using React, vanilla...?

**Reply by SemStassen** [reply]
@davidjerleke Ofcourse! And thanks for the swift reply.

My goal is to have both carousel `A` follow `B` and `B` follow `A` in a react project. The use case is almost exactly the same as the @7iomka described earlier in this thread, with slides having the same width in both carousels.

Once again thanks!

**Reply by davidjerleke** [reply]
@SemStassen and a follow up question: Will both carousels have the same scroll snaps and same setup? If no, what do you mean by follow? Do you mean the exact scroll progress in percent?

**Reply by SemStassen** [reply]
@davidjerleke Yes! the only difference between the setup (besides the actual content) is the height. So scroll snap and setup will be the same aswell. 

I got an example working where the actual slide is synced between the two different carrousels, but could not figure out a way to sync their exact scrollprogress (which is what I want).

**Reply by davidjerleke** [reply]
@SemStassen and will both carousels be open for input (drag events and nav buttons)?

**Reply by SemStassen** [reply]
@davidjerleke Yes, functionally they will be exactly the same. Both scrollable (with drag) and controllable via nav buttons.

**Reply by SemStassen** [reply]
Looking further into it, using `location` instead of `scrollProgress` seems promising, but is still not working as I intend.

Interacting with the `dayCarousel` only visually updates the `hourCarousel` position whenever it internally thinks the `hourCarousel` is still scrolling. It also acts very jittery. Maybe there is something I'm missing in my implementation?

```js
  const [dayApi, setDayApi] = useState<CarouselApi>();
  const [hourApi, setHourApi] = useState<CarouselApi>();

  const debouncedOnScroll = useDebouncedCallback(
    () => {
      if (!dayApi || !hourApi) return;

      const dayEngine = dayApi.internalEngine();
      const hourEngine = hourApi.internalEngine();

      hourEngine.location.set(dayEngine.location.get());
    },
    20,
    {
      leading: true,
      maxWait: 50,
    },
  );

  useEffect(() => {
    if (!dayApi) return;

    dayApi.on("scroll", debouncedOnScroll);

    return () => {
      dayApi.off("scroll", debouncedOnScroll);
    };
  }, [dayApi, hourApi, debouncedOnScroll]);
```
^ The above example is just for testing out syncing carousel A with B, not both ways

**Comment by JeffreyChix** [comment]
Hey,

Not sure if anyone's still looking to get this working, but I was able to successfully sync two carousels using this approach.

```jsx
const [emblaOneRef, emblaOneApi] = useEmblaCarousel();
const [emblaTwoRef, emblaTwoApi] = useEmblaCarousel();

const syncCarousels = useCallback((mainApi, secondaryApi) => {
  if (!mainApi || !secondaryApi) return;
  const selectedIndex = mainApi.selectedScrollSnap();
  secondaryApi.scrollTo(selectedIndex);
}, []);

useEffect(() => {
  if (!emblaOneApi || !emblaTwoApi) return;

  emblaOneApi.on('select', () => syncCarousels(emblaOneApi, emblaTwoApi));
  emblaTwoApi.on('select', () => syncCarousels(emblaTwoApi, emblaOneApi));
}, [emblaOneApi, emblaTwoApi, syncCarousels]);
```

In my case, I’m using the Shadcn UI Carousel component, which is a wrapper for Embla Carousel. Here’s what my implementation looks like:

```tsx
const [carouselOneApi, setCarouselOneApi] = useState<CarouselApi>();
const [carouselTwoApi, setCarouselTwoApi] = useState<CarouselApi>();

const syncCarousels = useCallback(
  (mainApi: CarouselApi, secondaryApi: CarouselApi) => {
    if (!mainApi || !secondaryApi) return;
    const selectedIndex = mainApi.selectedScrollSnap();
    secondaryApi.scrollTo(selectedIndex);
  },
  []
);

useEffect(() => {
  if (!carouselOneApi || !carouselTwoApi) return;

  carouselOneApi.on('select', () => syncCarousels(carouselOneApi, carouselTwoApi));
  carouselTwoApi.on('select', () => syncCarousels(carouselTwoApi, carouselOneApi));

  return () => {
    carouselOneApi.off('select', () => syncCarousels(carouselOneApi, carouselTwoApi));
    carouselTwoApi.off('select', () => syncCarousels(carouselTwoApi, carouselOneApi));
  };
}, [carouselOneApi, carouselTwoApi, syncCarousels]);
```

This worked well for me!

**Reply by 7iomka** [reply]
@JeffreyChix 
One mistake I can see here - unsubscription wouldn't work here, save anonumous function as constant first, than use it if off method.
I didn't tried it yet, but the old question for this feature request was if the embla can handle inifinite-loops internally.
